### ISSUE TITLE
cursor moves to cmdline when searching for pattern (perceived as flicker) (solution: make ui_busy_start actually mean "busy"!)

### ISSUE THREAD

**Issue by Asheq** [neovim/neovim#25980]
### Problem

The cursor briefly moves to the commandline when searching for the next pattern using:

- `n`
- `N`
- `*`
- `#`

## In terminal nvim
In terminal nvim, this is perceived as a cursor flicker.

https://github.com/neovim/neovim/assets/4657140/2aa1a535-9109-42f5-9515-46b9b49fcc0b

## In neovide (with cursor animation)
In neovide with cursor animation enabled, this is perceived as the cursor jumping across the screen and then back.

https://github.com/neovim/neovim/assets/4657140/e6020b1b-9b37-49b8-bf4e-5df5c6623a5c

### Open neovim:
```
nvim --clean
```
or
```
neovide -- --clean
```

### Create new buffer in right window
```
:vnew
<C-w>l
```

### Setup buffer to contain `xyz` on 10,000,000 lines and `abc` on line 1 and line 5,000,000 
```
iabc<Esc>
oxyz<Esc>
yy10000000p
50%
Oabc<Esc>
```

### Invoke search on `abc`:
```
*
n
N
#
```

Notice that cursor goes to commandline between searches.

### Expected behavior

The cursor should not move to the commandline.

It should stay in the same screen position until the next search match location is found and then move there directly.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-1461+g468292dcb-Homebrew

### Vim (not Nvim) behaves the same?

yes, vim 9.0.1544 (cursor disappears briefly as well but does not seem to appear in comandline)

### Operating system/version

macOS 13.5.1

### Terminal name/version

kitty 0.29.0

### $TERM environment variable

xterm-kitty

### Installation

homebrew

**Comment by Asheq** [comment]
PS: The reason that I created this as a separate issue from #25974 is because the cursor moves to the commandline for a different reason: in that issue it's because a message is being printing, whereas in this issue it appears to be because neovim is "resting" or "parking" the cursor in the commandline until it finds the next match.

**Comment by Asheq** [comment]
The downstream issue in neovide: https://github.com/neovide/neovide/issues/1665

**Comment by zeertzjq** [comment]
This seem intentional. There is even an explicit cursor move to cmdline:
https://github.com/neovim/neovim/blob/fdaf6bc5573f578d100321dfc084d513884fd73a/src/nvim/search.c#L1271

**Comment by Asheq** [comment]
If there's no good reason for the explicit cursor move, I would favor removing it. It does cause a bad user experience, especially with an animated cursor as shown in the screen capture above, and in the downstream issue in neovide.

I'm not familiar with C or the code-base at all, but I imagine that the reasoning for moving the cursor to the command line was to put it in a temporary "holding area" until the next match was found. A better UX would be to just leave the cursor exactly where it is. Then when it is finally moved to the next match, the transition/animation is natural.

**Comment by justinmk** [comment]
This is a feature, it's how vim indicates "busy" state.

However, [the code](https://github.com/neovim/neovim/blob/b923fcbaf06276051372c17177a9970c4adc8e3d/src/nvim/search.c#L1311) that @zeertzjq mentioned can be improved, it should call `ui_busy_start` / `ui_busy_stop`, not a literal call to `gotocmdline`. Then UIs can choose how to signal "busy".

**Comment by Asheq** [comment]
> The call to `gotocmdline` is probably still needed

Why is that?

> but signaling "busy" will allow UIs to opt-out of showing the cursor.

It would be great it if a UI could opt to continue showing the cursor during "busy" but leave it where it is for the sake of natural-looking cursor animations.

**Comment by Shougo** [comment]
> Why is that?

This is to let the user know that processing is taking place.
The cursor being on the command line indicates that the user is not available to operate.
This is equivalent to the hourglass icon in Windows GUI.

**Comment by Shougo** [comment]
@Asheq You can test #29446.

**Comment by Asheq** [comment]
> @Asheq You can test #29446.

Thanks. The cursor no longer appears to go the command line during search.

It does disappear and then reappear in the new location, though. This is fine.

**Comment by fredizzimo** [comment]
I'm trying to test this in Neovide, but I have problems with this part of the design

>>> The call to gotocmdline is probably still needed

>> Why is that?

> This is to let the user know that processing is taking place.
The cursor being on the command line indicates that the user is not available to operate.
This is equivalent to the hourglass icon in Windows GUI.

`ui_busy_start` is already in indication of that, why do we need two ways of indicating the same thing? An UI showing a hourglass does not need the cursor to move at all.

With https://github.com/neovim/neovim/pull/29446, we now receive the following events:
```
busy_start
grid_line
grid_cursor_goto { grid: 4, row: 0, column: 0 }
flush
busy_stop
grid_line
grid_line
grid_line
win_viewport
grid_cursor_goto { grid: 2, row: 12, column: 39 }
flush
```
Which means, that even if we hide the cursor on `busy_start`, there will be an animation from the command line to the new positions, which just looks wrong, it should animated from the cursor position that was set before the busy position to the new one.

We could ignore all `grid_cursor_goto` events during the busy period, but that opens a whole category of potential bugs, where the cursor is moved during `busy_start` but never restored. 

The TUI works fiine with https://github.com/neovim/neovim/pull/29446, since it hides the cursor totally, and therefore it does not need the cursor movement.

**Comment by justinmk** [comment]
@fredizzimo what do you suggest? We could move the `gotocmdline` decision into the TUI (that would be how it handles "busy"). But I can't tell if you're saying the current "busy on/off" signals are already sufficient for UIs to handle.

**Comment by fredizzimo** [comment]
Yes, I'm suggesting that `ui_busy_start` should be enough for all UI types, both TUI and other UIs. But in practice it might be a breaking change for some UI, especially since the documentation says this:
> Indicates to the UI that it must stop rendering the cursor. This event is misnamed and does not actually have anything to do with busyness.

I was also thinking about a potential hack that we could use in Neovide, which allows `cursor_goto` during the busy periods. It's not elegant, but has the advantage that it handles other types of busy periods as well, and there's a smaller chance of something going wrong.  But I think it's certainly better than our [current hack](https://github.com/neovide/neovide/pull/753), which causes some bugs on its own, and would therefore like to be able to remove when Neovim 0.11 is released.

* If a `grid_cursor_goto` is received during a UI batch where the cursor is disabled, set `desired_cursor_target` to that position, but don't update the actual (hidden) cursor position
* When `nvim_busy_end` is received animated from the current cursor position, the one before `ui_busy_start` was called to `desired_cursor_target`

**Comment by fredizzimo** [comment]
Actually `must stop rendering the cursor` implies that extra `grid_cursor_goto` during the busy period should not be needed and therefore is not a breaking change.

But still, it would be nice if it actually signals a busy state, so that the UI can indicate it as fits.

**Comment by justinmk** [comment]
> This event (`busy_start`) is misnamed and does not actually have anything to do with busyness.

Forgot about this. This came from https://github.com/neovim/neovim/pull/14415 but I can't find details, and most of the code IMO looks like it could reasonably be defined as "busyness". @glacambre @bfredl do you see risk in revisiting this? Do we have another way to signal "busyness" ?

**Comment by fredizzimo** [comment]
It looks like the request came from this comment
* https://github.com/neovim/neovim/issues/14386#issuecomment-823067514

**Comment by glacambre** [comment]
>do you see risk in revisiting this?

I guess the problem is that `busy_start` is sent when entering terminal mode, and any "busyness" indicator (like turning the mouse cursor into an hourglass) while in terminal mode would look weird.

To my knowledge, there is no way for Neovim to signal that it is stuck processing something.

**Comment by justinmk** [comment]
> the problem is that `busy_start` is sent when entering terminal mode

Ah, in that case this would be resolved by https://github.com/neovim/neovim/pull/26835 . Then we could reclaim `ui_busy_start` to really mean "busy".

**Comment by fredizzimo** [comment]
https://github.com/neovim/neovim/pull/31562 (which replaced https://github.com/neovim/neovim/pull/26835) has been merged now, so I guess the final blocker is removed.

**Comment by fredizzimo** [comment]
What's holding these two up now?
* https://github.com/neovim/neovim/pull/29446
* https://github.com/neovim/neovim/pull/29447

**Comment by fredizzimo** [comment]
Does the TUI really need something more? It already hides the cursor with busy start, which IMO is a stronger indication of busyness than moving the cursor there.

Neovide would love to get those in for 0.11, and possible even some more cases after some real-world testing without our hacks activated.

**Comment by fredizzimo** [comment]
Since 0.11 is out, and we have two pull requests for that fixes this problem without any issues as far as I can see. I would like to see them merged as soon as possible. Or at least get a plan what needs to be done. 

Those two PRs fixes at least 5 different bugs where the cursor is completely in the wrong place in Neovide, which is a result of trying to do a very dirty hack to work around the flickering problem, which is very severe in in Neovide without it. It's possible that those two PRs are not enough, but we don't know unless we have a wide real-world testing, and by merging earlier we have a longer time to test until the 0.12 release.

**Comment by justinmk** [comment]
> What's holding these two up now?

at least one of them is failing every ci job?
